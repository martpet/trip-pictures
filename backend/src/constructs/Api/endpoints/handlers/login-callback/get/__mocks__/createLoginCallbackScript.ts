export const createLoginCallbackScript = vi
  .fn()
  .mockName('createLoginCallbackScript')
  .mockReturnValue({
    script: 'dummyScript',
    cspHash: 'dummyCspHash',
  });
