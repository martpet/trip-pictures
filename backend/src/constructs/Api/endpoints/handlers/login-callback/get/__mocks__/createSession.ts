export const createSession = vi
  .fn()
  .mockName('createSession')
  .mockResolvedValue('dummySessionCookie');
