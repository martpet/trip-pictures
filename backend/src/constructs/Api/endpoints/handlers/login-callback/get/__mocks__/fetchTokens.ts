import { IdTokenPayload, OauthTokens } from '~/constructs/Api/types';

export const dummyOauthTokens: OauthTokens = {
  accessToken: 'dummyAccessToken',
  refreshToken: 'dummyRefreshToken',
  idToken: 'dummyIdToken',
  expiresIn: 123,
  idTokenPayload: {
    sub: 'dummySub',
    nonce: 'dummyIdTokenNonce',
    aud: 'dummyAud',
  } as IdTokenPayload,
};

export const fetchTokens = vi
  .fn()
  .mockName('fetchTokens')
  .mockResolvedValue(dummyOauthTokens);
