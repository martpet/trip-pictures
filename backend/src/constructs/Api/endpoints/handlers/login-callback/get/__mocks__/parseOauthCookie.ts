import { OauthCookieProps } from '~/constructs/Api/types';

export const dummyOauthCookieProps: OauthCookieProps = {
  stateNonce: 'dummyState',
  idTokenNonce: 'dummyIdTokenNonce',
  codeVerifier: 'dummyCodeVerifier',
};

export const parseOauthCookie = vi
  .fn()
  .mockName('parseOauthCookie')
  .mockReturnValue(dummyOauthCookieProps);
