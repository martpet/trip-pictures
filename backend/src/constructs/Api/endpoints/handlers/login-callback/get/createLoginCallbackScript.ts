import { createSha256CspHash, EnvName, loginWindowSuccessData } from 'lambda-layer';

type CreateLoginCallbackScriptProps = {
  envName: EnvName;
  appDomain: string;
};

export const createLoginCallbackScript = ({
  envName,
  appDomain,
}: CreateLoginCallbackScriptProps) => {
  const targetOrigin = envName === 'personal' ? '*' : `https://${appDomain}`;
  const script = `opener.postMessage("${loginWindowSuccessData}", "${targetOrigin}")`;

  return { script, cspHash: createSha256CspHash(script) };
};
