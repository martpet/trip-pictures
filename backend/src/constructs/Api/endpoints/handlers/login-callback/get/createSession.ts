import {
  cookie,
  cookieName,
  crypto,
  DynamoDBClient,
  DynamoDBDocumentClient,
  millis,
  OauthTokens,
  PutCommand,
  refreshTokenValidityInDays,
  SessionsTableItem,
  sessionsTableOptions,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

type CreateSessionProps = {
  tokens: OauthTokens;
};

export const createSession = async ({ tokens }: CreateSessionProps) => {
  const { idTokenPayload, refreshToken, idToken } = tokens;
  const { envName } = globalLambdaProps;
  const sessionId = crypto.randomUUID();
  const created = Date.now();
  const refreshTokenExpires = (created + millis.days(refreshTokenValidityInDays)) / 1000;

  const putCommand = new PutCommand({
    TableName: sessionsTableOptions.tableName,
    Item: {
      id: sessionId,
      userId: idTokenPayload.sub,
      created,
      refreshToken,
      refreshTokenExpires,
      idToken,
    } satisfies SessionsTableItem,
  });

  await ddbDocClient.send(putCommand);

  const sessionCookie = cookie.serialize(cookieName('sessionId'), sessionId, {
    secure: true,
    httpOnly: true,
    sameSite: envName !== 'personal' || 'none',
    expires: new Date(refreshTokenExpires * 1000),
  });

  return sessionCookie;
};
