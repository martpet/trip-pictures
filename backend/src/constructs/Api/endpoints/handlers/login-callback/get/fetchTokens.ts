import camelcaseKeys from 'camelcase-keys';
import {
  authPaths,
  FetchTokensResponse,
  getIdTokenPayload,
  OauthTokens,
} from 'lambda-layer';
import fetch from 'node-fetch';

type FetchTokensProps = {
  code: string;
  codeVerifier: string;
  clientId: string;
  authDomain: string;
  loginCallbackUrl: string;
};

export const fetchTokens = async ({
  code,
  codeVerifier,
  clientId,
  authDomain,
  loginCallbackUrl,
}: FetchTokensProps): Promise<OauthTokens> => {
  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'authorization_code',
      client_id: clientId,
      code,
      code_verifier: codeVerifier,
      redirect_uri: loginCallbackUrl,
    }),
  });

  const data = (await response.json()) as FetchTokensResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  return {
    ...camelcaseKeys(data),
    idTokenPayload: await getIdTokenPayload(data.id_token),
  };
};
