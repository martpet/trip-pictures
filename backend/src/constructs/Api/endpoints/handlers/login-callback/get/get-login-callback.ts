import {
  APIGatewayProxyHandlerV2,
  ApiQueryStrings,
  appEnvs,
  EnvVars,
  errorResponse,
  StatusCodes,
} from 'lambda-layer';

import { createLoginCallbackScript } from './createLoginCallbackScript';
import { createSession } from './createSession';
import { fetchTokens } from './fetchTokens';
import { parseOauthCookie } from './parseOauthCookie';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { envName } = globalLambdaProps;
  const { appDomain } = appEnvs[envName];
  const { script: responseScript } = createLoginCallbackScript({ envName, appDomain });

  const {
    state,
    code,
    error: queryStringError,
    error_description: queryStringErrorDescrption,
  } = Object(event.queryStringParameters) as ApiQueryStrings<'/login-callback'>;

  const { clientId, authDomain, loginCallbackUrl } = process.env as EnvVars<
    '/login-callback',
    'GET'
  >;

  const { stateNonce, idTokenNonce, codeVerifier } = parseOauthCookie(event);

  if (queryStringError) {
    return errorResponse('5vR4w3wVJs', {
      error: `${queryStringError}: ${queryStringErrorDescrption}`,
    });
  }

  if (!state || !code) {
    return errorResponse('1cByWITGGHw', { statusCode: StatusCodes.BAD_REQUEST });
  }

  if (!clientId || !authDomain || !loginCallbackUrl) {
    return errorResponse('XuJRO7_43O');
  }

  if (!stateNonce || !idTokenNonce || !codeVerifier) {
    return errorResponse('OSUiwUI_DG');
  }

  if (state !== stateNonce) {
    return errorResponse('eWe87M5edc', { statusCode: StatusCodes.UNAUTHORIZED });
  }

  const tokens = await fetchTokens({
    code,
    codeVerifier,
    clientId,
    authDomain,
    loginCallbackUrl,
  });

  if (tokens.idTokenPayload.nonce !== idTokenNonce) {
    return errorResponse('C25fjaal7o', { statusCode: StatusCodes.UNAUTHORIZED });
  }

  const sessionCookie = await createSession({ tokens });

  return {
    statusCode: StatusCodes.OK,
    cookies: [sessionCookie],
    headers: { 'Content-Type': 'text/html' },
    body: `<script>${responseScript}</script>`, // [todo] Add "Please wait" visuals
  };
};
