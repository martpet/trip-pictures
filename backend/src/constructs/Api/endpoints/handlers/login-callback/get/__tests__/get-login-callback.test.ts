import { describe } from 'vitest';

import { OauthCookieProps } from '~/constructs/Api/types';
import {
  itCalls,
  itHasEnvVars,
  itHasQueryStrings,
  itResolves,
  itResolvesWithError,
} from '~/constructs/Api/utils';
import { capitalize } from '~/utils';

import { dummyOauthTokens } from '../__mocks__/fetchTokens';
import { dummyOauthCookieProps } from '../__mocks__/parseOauthCookie';
import { createLoginCallbackScript } from '../createLoginCallbackScript';
import { createSession } from '../createSession';
import { fetchTokens } from '../fetchTokens';
import { handler } from '../get-login-callback';
import { parseOauthCookie } from '../parseOauthCookie';

vi.importMock('../createSession');

vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('../fetchTokens');
vi.mock('../parseOauthCookie');
vi.mock('../createSession');
vi.mock('../createLoginCallbackScript');

process.env.authDomain = 'dummyAuthDomain';
process.env.clientId = 'dummyClientId';
process.env.loginCallbackUrl = 'dummyLoginCallbackUrl';
process.env.envName = 'dummyEnvName';

const args = [
  {
    cookies: 'dummyCookies',
    queryStringParameters: {
      code: 'dummyCode',
      state: 'dummyState',
      error_description: 'dummyQueryStringErrorDescription',
    },
  },
] as unknown as Parameters<typeof handler>;

describe('"get-login-callback" handler', () => {
  itHasQueryStrings(['code', 'state'], handler, args);
  itHasEnvVars(['clientId', 'authDomain', 'loginCallbackUrl'], handler, args);
  itCalls(parseOauthCookie, handler, args);
  itCalls(fetchTokens, handler, args);
  itCalls(createSession, handler, args);
  itCalls(createLoginCallbackScript, handler, args);
  itResolves(handler, args);

  describe.each(['error'])('when "%s" query string parameter is present', (key) => {
    const argsClone = structuredClone(args);
    (argsClone[0] as Required<(typeof argsClone)[0]>).queryStringParameters[
      key
    ] = `dummyQueryString${capitalize(key)}`;
    itResolvesWithError(handler, argsClone);
  });

  describe.each(['stateNonce', 'idTokenNonce', 'codeVerifier'])(
    'when "oauth" cookie content is missing "%s" key',
    (key) => {
      const oauthCookiePropsClone = structuredClone(dummyOauthCookieProps);
      delete oauthCookiePropsClone[key as keyof OauthCookieProps];
      beforeEach(() => {
        vi.mocked(parseOauthCookie).mockReturnValueOnce(oauthCookiePropsClone);
      });
      itResolvesWithError(handler, args);
    }
  );

  describe('when "stateNonce" from "oauth" cookie does not match with "state" from query strings params', () => {
    const argsClone = structuredClone(args);
    const { queryStringParameters } = argsClone[0] as Required<(typeof argsClone)[0]>;
    queryStringParameters.state = 'differentState';
    itResolvesWithError(handler, argsClone);
  });

  describe('when "idTokenNonce" from "oauth" cookie does not match with "nonce" from the "idToken" payload', () => {
    const tokensPropsClone = structuredClone(dummyOauthTokens);
    tokensPropsClone.idTokenPayload.nonce = 'differentNonce';
    beforeEach(() => {
      vi.mocked(fetchTokens).mockResolvedValueOnce(tokensPropsClone);
    });
    itResolvesWithError(handler, args);
  });
});
