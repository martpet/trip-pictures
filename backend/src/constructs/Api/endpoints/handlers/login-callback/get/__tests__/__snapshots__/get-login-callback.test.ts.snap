// Vitest Snapshot v1

exports[`"get-login-callback" handler > calls "createLoginCallbackScript" with correct args 1`] = `
[
  [
    {
      "appDomain": "trip.pictures",
      "envName": "production",
    },
  ],
]
`;

exports[`"get-login-callback" handler > calls "createSession" with correct args 1`] = `
[
  [
    {
      "tokens": {
        "accessToken": "dummyAccessToken",
        "expiresIn": 123,
        "idToken": "dummyIdToken",
        "idTokenPayload": {
          "aud": "dummyAud",
          "nonce": "dummyIdTokenNonce",
          "sub": "dummySub",
        },
        "refreshToken": "dummyRefreshToken",
      },
    },
  ],
]
`;

exports[`"get-login-callback" handler > calls "fetchTokens" with correct args 1`] = `
[
  [
    {
      "authDomain": "dummyAuthDomain",
      "clientId": "dummyClientId",
      "code": "dummyCode",
      "codeVerifier": "dummyCodeVerifier",
      "loginCallbackUrl": "dummyLoginCallbackUrl",
    },
  ],
]
`;

exports[`"get-login-callback" handler > calls "parseOauthCookie" with correct args 1`] = `
[
  [
    {
      "cookies": "dummyCookies",
      "queryStringParameters": {
        "code": "dummyCode",
        "error_description": "dummyQueryStringErrorDescription",
        "state": "dummyState",
      },
    },
  ],
]
`;

exports[`"get-login-callback" handler > resolves with a correct value 1`] = `
{
  "body": "<script>dummyScript</script>",
  "cookies": [
    "dummySessionCookie",
  ],
  "headers": {
    "Content-Type": "text/html",
  },
  "statusCode": 200,
}
`;

exports[`"get-login-callback" handler > when "authDomain" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "XuJRO7_43O",
  ],
]
`;

exports[`"get-login-callback" handler > when "authDomain" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "clientId" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "XuJRO7_43O",
  ],
]
`;

exports[`"get-login-callback" handler > when "clientId" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "code" query string parameter is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1cByWITGGHw",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`"get-login-callback" handler > when "code" query string parameter is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "error" query string parameter is present > calls "errorResponse" with correct args 1`] = `
[
  [
    "5vR4w3wVJs",
    {
      "error": "dummyQueryStringError: dummyQueryStringErrorDescription",
    },
  ],
]
`;

exports[`"get-login-callback" handler > when "error" query string parameter is present > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "idTokenNonce" from "oauth" cookie does not match with "nonce" from the "idToken" payload > calls "errorResponse" with correct args 1`] = `
[
  [
    "C25fjaal7o",
    {
      "statusCode": 401,
    },
  ],
]
`;

exports[`"get-login-callback" handler > when "idTokenNonce" from "oauth" cookie does not match with "nonce" from the "idToken" payload > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "loginCallbackUrl" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "XuJRO7_43O",
  ],
]
`;

exports[`"get-login-callback" handler > when "loginCallbackUrl" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "oauth" cookie content is missing "codeVerifier" key > calls "errorResponse" with correct args 1`] = `
[
  [
    "OSUiwUI_DG",
  ],
]
`;

exports[`"get-login-callback" handler > when "oauth" cookie content is missing "codeVerifier" key > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "oauth" cookie content is missing "idTokenNonce" key > calls "errorResponse" with correct args 1`] = `
[
  [
    "OSUiwUI_DG",
  ],
]
`;

exports[`"get-login-callback" handler > when "oauth" cookie content is missing "idTokenNonce" key > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "oauth" cookie content is missing "stateNonce" key > calls "errorResponse" with correct args 1`] = `
[
  [
    "OSUiwUI_DG",
  ],
]
`;

exports[`"get-login-callback" handler > when "oauth" cookie content is missing "stateNonce" key > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "state" query string parameter is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1cByWITGGHw",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`"get-login-callback" handler > when "state" query string parameter is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login-callback" handler > when "stateNonce" from "oauth" cookie does not match with "state" from query strings params > calls "errorResponse" with correct args 1`] = `
[
  [
    "eWe87M5edc",
    {
      "statusCode": 401,
    },
  ],
]
`;

exports[`"get-login-callback" handler > when "stateNonce" from "oauth" cookie does not match with "state" from query strings params > resolves with a correct value 1`] = `"dummyErrorResponse"`;
